python-package/xgboost/spark/core.py

```diff
@@ -0,0 +1,877 @@
+            get_logger(self.__class__.__name__).warning(
+                f"The num_workers {num_workers} set for xgboost distributed "
+                f"training is greater than current max number of concurrent "
+                f"spark task slots, you need wait until more task slots available "
+                f"or you need increase spark cluster workers."
+            )
+
+        if self._repartition_needed(dataset):
+            dataset = dataset.withColumn(
+                "values", col("values").cast(ArrayType(FloatType()))
+            )
+            dataset = dataset.repartition(num_workers)
+        train_params = self._get_distributed_config(dataset, fit_params)
+
+        def _train_booster(pandas_df_iter):
+            """
+            Takes in an RDD partition and outputs a booster for that partition after going through
+            the Rabit Ring protocol
+            """
+            from pyspark import BarrierTaskContext
+
+            context = BarrierTaskContext.get()
+
+            dtrain, dval = None, []
+            if has_validation:
+                dtrain, dval = convert_partition_data_to_dmatrix(
+                    pandas_df_iter, has_weight, has_validation
+                )
+                dval = [(dtrain, "training"), (dval, "validation")]
+            else:
+                dtrain = convert_partition_data_to_dmatrix(
+                    pandas_df_iter, has_weight, has_validation
+                )
+
+            booster_params, kwargs_params = self._get_dist_booster_params(train_params)
+            context.barrier()
+            _rabit_args = ""
+            if context.partitionId() == 0:
+                _rabit_args = str(_get_rabit_args(context, num_workers))
+
+            messages = context.allGather(message=str(_rabit_args))
+            _rabit_args = _get_args_from_message_list(messages)
+            evals_result = {}
+            with RabitContext(_rabit_args, context):
+                booster = worker_train(
+                    params=booster_params,
+                    dtrain=dtrain,
+                    evals=dval,
+                    evals_result=evals_result,
+                    **kwargs_params,
+                )
+            context.barrier()
+
+            if context.partitionId() == 0:
+                yield pd.DataFrame(data={"booster_bytes": [cloudpickle.dumps(booster)]})
+
+        result_ser_booster = (
+            dataset.mapInPandas(_train_booster, schema="booster_bytes binary")
+            .rdd.barrier()
+            .mapPartitions(lambda x: x)
+            .collect()[0][0]
+        )
+        result_xgb_model = self._convert_to_model(cloudpickle.loads(result_ser_booster))
+        return self._copyValues(self._create_pyspark_model(result_xgb_model))
+
+    def _fit(self, dataset):
+        self._validate_params()
+        # Unwrap the VectorUDT type column "feature" to 4 primitive columns:
+        # ['features.type', 'features.size', 'features.indices', 'features.values']
```

Comment:
Looks like vector_to_array just converts the Vector to Arrays, so the comments here seem not to be used any more?

Reply:
This comment comes from supporting pyspark Sparse Vector (it is not included in this PR for now, but we will add it later), not relates to this PR. I will remove it.